pkgs/applications/blockchains/bisq-desktop/mkdeps.nix

```diff
@@ -0,0 +1,103 @@
+# to provide the (Java) dependencies so that the package can be
+# built in a reproducable way.
+
+# USAGE:
+# 1. `nix build -f mkdeps.nix`
+# If needed, update the `outputHash` at REF1, below. Then repeat step 1.
+# 2. `./result > deps.nix`
+# 3. `cd <nixpkgs repo root>`
+# 4. Build as usual: `nix build bisq-desktop`
+
+with import <nixpkgs> { };
+let
+  common = callPackage ./common.nix { };
+  inherit (common) src version pname grpc gradle;
+
+  # fake build to pre-download deps into fixed-output derivation
+  prebuild = pkgs.stdenv.mkDerivation {
+    name = "${pname}-prebuild-${version}";
+    inherit src;
+    buildInputs = [ gradle pkgs.perl pkgs.unzip pkgs.zip ];
+
+    patchPhase = ''
+      substituteInPlace ./build.gradle \
+        --replace 'artifact = "com.google.protobuf:protoc:''${protocVersion}"' "path = '${pkgs.protobuf3_10}/bin/protoc'" \
+        --replace 'artifact = "io.grpc:protoc-gen-grpc-java:''${grpcVersion}"' "path = '${grpc}/bin/protoc-gen-rpc-java'"
+    '';
+
+    buildPhase = ''
+      export GRADLE_USER_HOME=$(mktemp -d)
+      gradle --no-daemon desktop:build --exclude-task desktop:test
+    '';
+
+    # perl code mavenizes pathes (com.squareup.okio/okio/1.13.0/a9283170b7305c8d92d25aff02a6ab7e45d06cbe/okio-1.13.0.jar -> com/squareup/okio/okio/1.13.0/okio-1.13.0.jar)
+    installPhase = ''
+      find $GRADLE_USER_HOME -type f -regex '.*\.\(jar\|pom\)' \
+        | perl -pe 's#(.*/([^/]+)/([^/]+)/([^/]+)/[0-9a-f]{30,40}/([^/\s]+))$# ($x = $2) =~ tr|\.|/|; "install -Dm444 $1 \$out/$x/$3/$4/$5" #e' \
+        | sh
+    '';
+
+    outputHashAlgo = "sha256";
+    outputHashMode = "recursive";
+
+    # REF1
+    outputHash = "1zx6hd7b4kc2s4zv9p897dbsydvh260hdhhwqpaliyq300vhfiwv";
+  };
+
+  gen-deps-script = pkgs.writeScript "${pname}-gen-deps-script" ''
+    echo "# This file was generated with mkdeps.nix"
+    echo "# You should not edit this file manually."
+    echo
+    echo "["
+
+    for path in $(find -L ${prebuild} -type f | sort | grep -v gradle-witness); do
+      name=$(basename $path)
+      sha256=""
+      upstreamPath=$(realpath --relative-to ${prebuild} $path)
+      upstreamDir=$(dirname $upstreamPath)
+
+      for repo in https://repo.maven.apache.org/maven2 https://jitpack.io https://jcenter.bintray.com; do
+        url="$repo/$upstreamPath"
+        sha256=$(${pkgs.nix}/bin/nix-prefetch-url $url 2> /dev/null)
+        status=$?
+
+        if [[ $status -eq 0 ]]
+        then
+          break
+        fi
+      done
+
+      if [[ "$sha256" == "" ]]
+      then
+        echo "ERROR: Unable to download $upstreamPath from any known sources."
+        exit 1
+      fi
+
+      echo "  {"
+      echo "    url = \"$url\";"
+      echo "    sha256 = \"$sha256\";"
+      echo "    name = \"$name\";"
+      echo "    mavenDir = \"$upstreamDir\";"
+      echo "  }"
+    done
+
+    echo "]"
+  '';
+
+in
+pkgs.stdenv.mkDerivation {
+  name = "${pname}-deps-${version}.nix";
```

Comment:
If we hardcode the version the deps need to be rebuild for every update even if they stay the same.